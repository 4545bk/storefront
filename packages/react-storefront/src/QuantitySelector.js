/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Fragment, Component } from 'react'
import Add from '@material-ui/icons/Add'
import Remove from '@material-ui/icons/Remove'
import IconButton from '@material-ui/core/IconButton'
import Input from '@material-ui/core/Input'
import withStyles from '@material-ui/core/styles/withStyles'
import PropTypes from 'prop-types'
import { observer, inject } from 'mobx-react'

export const styles = theme => ({
  root: {
    backgroundColor: theme.palette.divider,
    border: 'none',
    width: '110px',
    padding: 0
  },
  icon: {
    fontSize: theme.typography.title.fontSize,
    top: '-4px',
    position: 'relative'
  },
  button: {
    height: '36px',
    width: '36px'
  },
  input: {
    color: theme.typography.body1.color,
    textAlign: 'center',
    padding: 0
  },
  focused: {
    backgroundColor: theme.palette.divider
  },
  underline: {
    '&::before': {
      display: 'none'
    }
  },
})

/**
 * A quantity selector with plus and minus buttons. Any extra props are spread to the 
 * underlying Material UI input element.
 */
@inject(({ app, ampStateId }) => ({ app, ampStateId }))
@withStyles(styles, { name: 'RSFQuantitySelector' })
@observer
export default class QuantitySelector extends Component {

  static propTypes = {
    /**
     * CSS classes
     */
    classes: PropTypes.object,

    /**
     * The plus icon
     */
    addIcon: PropTypes.element,

    /**
     * The minus icon
     */
    subtractIcon: PropTypes.element,

    /**
     * The current value
     */
    value: PropTypes.number,

    /**
     * The minimum value.  Defaults to 1.
     */
    minValue: PropTypes.number,

    /**
     * The maximum value.  Defaults to 100.
     */
    maxValue: PropTypes.number,

    /**
     * Called when the value is changed.  The new value is passed as the only argument
     */
    onChange: PropTypes.func,

    /**
     * If specified, this component will automatically control the price of a product.
     * This should be an instance of ProductModelBase
     */
    product: PropTypes.object
  }

  static defaultProps = {
    onChange: Function.prototype,
    minValue: 1,
    maxValue: 100,
    value: 1
  }

  render() {
    let { 
      app,
      classes, 
      addIcon, 
      subtractIcon, 
      value, 
      minValue,
      maxValue,
      analytics,
      product,
      inputProps,
      ampStateId,
      ...other 
    } = this.props

    if (product) {
      value = product.quantity
    }

    const { quantitySelector, icon, button, ...inputClasses } = classes

    const bindProps = {
      inputProps: {
        ...inputProps,
        "amp-bind": `value=>${ampStateId}.quantity`
      },
      [app.amp ? 'readOnly' : 'disabled']: true
    }

    return (
      <Fragment>
        <Input
          startAdornment={
            <IconButton 
              size="small" 
              classes={{ root: button }} 
              onClick={() => this.onChange(value - 1)} 
              on={`tap:AMP.setState({ ${ampStateId}: { quantity: max(${minValue}, (${ampStateId}.quantity || ${value}) - 1) } })`}
            >
              {subtractIcon || <Remove classes={{ root: icon }}/>}
            </IconButton>
          }
          endAdornment={
            <IconButton 
              size="small" 
              classes={{ root: button }} 
              onClick={() => this.onChange(value + 1)}
              on={`tap:AMP.setState({ ${ampStateId}: { quantity: min(${maxValue}, (${ampStateId}.quantity || ${value}) + 1) } })`}
            >
              {addIcon || <Add classes={{ root: icon }}/>}
            </IconButton>
          }
          onChange={this.onChange}
          value={value}
          classes={{
            underline: classes.underline,
            ...inputClasses
          }}
          {...bindProps}
          {...other}
        />
      </Fragment>
    )
  }

  onChange = value => {
    const { minValue, maxValue, product } = this.props

    if (value >= minValue && value <= maxValue) {
      this.props.onChange(value)
      
      if (product) {
        product.setQuantity(value)
      }
    }
  }

}
