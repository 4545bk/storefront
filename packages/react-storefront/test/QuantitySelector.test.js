/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import QuantitySelector from '../src/QuantitySelector'
import { mount } from 'enzyme'
import { Provider } from "mobx-react"
import AppModelBase from '../src/model/AppModelBase'

describe('QuantitySelector', () => {

  let app

  beforeEach(() => {
    app = AppModelBase.create
  })

  it('renders', () => {
    const component = (
      <Provider app={app}>
        <QuantitySelector/>
      </Provider>
    )

    expect(mount(component)).toMatchSnapshot()
  })

  it('should accept a value', () => {
    const component = (
      <Provider app={app}>
        <QuantitySelector value={2}/>
      </Provider>
    )
    expect(mount(component)).toMatchSnapshot()
  })

  it('should decrement when the minus button is clicked', () => {
    const onChange = jest.fn()

    mount(
      <Provider app={app}>
        <QuantitySelector value={2} onChange={onChange}/>
      </Provider>
    )
      .find('button').at(0)
      .simulate('click')

    expect(onChange).toBeCalledWith(1)
  })

  it('should enforce the minValue', () => {
    const onChange = jest.fn()

    mount(
      <Provider app={app}>
        <QuantitySelector value={0} minValue={0} onChange={onChange}/>
      </Provider>
    )
      .find('button').at(0)
      .simulate('click')

    expect(onChange).not.toBeCalled()
  })

  it('should enforce the maxValue', () => {
    const onChange = jest.fn()

    mount(
      <Provider app={app}>
        <QuantitySelector value={10} maxValue={10} onChange={onChange}/>
      </Provider>
    )
      .find('button').at(1)
      .simulate('click')

    expect(onChange).not.toBeCalled()
  })

  it('should increment when the plus button is clicked', () => {
    const onChange = jest.fn()
    
    mount(
      <Provider app={app}>
        <QuantitySelector value={1} onChange={onChange}/>
      </Provider>
    )
      .find('button').at(1)
      .simulate('click')

    expect(onChange).toBeCalledWith(2)
  })

})