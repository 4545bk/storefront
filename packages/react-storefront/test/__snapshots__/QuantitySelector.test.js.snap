// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantitySelector renders 1`] = `
<Provider app={[Function: res]}>
  <inject-WithStyles(QuantitySelector)>
    <WithStyles(QuantitySelector) app={[Function: res]} ampStateId={[undefined]}>
      <QuantitySelector classes={{...}} app={[Function: res]} ampStateId={[undefined]} onChange={[Function]} minValue={1} maxValue={100} value={1}>
        <WithStyles(Input) startAdornment={{...}} endAdornment={{...}} onChange={[Function]} value={1} classes={{...}} inputProps={{...}} disabled={true}>
          <Input classes={{...}} startAdornment={{...}} endAdornment={{...}} onChange={[Function]} value={1} inputProps={{...}} disabled={true} disableUnderline={false} fullWidth={false} multiline={false} type="text">
            <div className="MuiInput-root-7 RSFQuantitySelector-root-1 MuiInput-disabled-10 MuiInput-underline-11 RSFQuantitySelector-underline-6">
              <WithStyles(IconButton) size="small" classes={{...}} onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 1) - 1) } })">
                <IconButton classes={{...}} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 1) - 1) } })" color="default" disabled={false}>
                  <WithStyles(ButtonBase) className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 1) - 1) } })">
                    <ButtonBase classes={{...}} className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 1) - 1) } })" disableRipple={false} tabIndex="0" type="button">
                      <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-26 MuiIconButton-root-20 RSFQuantitySelector-button-3" type="button" disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 1) - 1) } })">
                        <span className="MuiIconButton-label-25">
                          <pure(Remove) classes={{...}}>
                            <Remove classes={{...}}>
                              <WithStyles(SvgIcon) classes={{...}}>
                                <SvgIcon classes={{...}} color="inherit" viewBox="0 0 24 24">
                                  <svg className="MuiSvgIcon-root-29 RSFQuantitySelector-icon-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true">
                                    <g>
                                      <path d="M19 13H5v-2h14v2z" />
                                    </g>
                                  </svg>
                                </SvgIcon>
                              </WithStyles(SvgIcon)>
                            </Remove>
                          </pure(Remove)>
                        </span>
                        <WithStyles(TouchRipple) innerRef={[Function: value]} center={true}>
                          <TouchRipple classes={{...}} center={true}>
                            <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-35" childFactory={[Function: childFactory]}>
                              <span className="MuiTouchRipple-root-35" />
                            </TransitionGroup>
                          </TouchRipple>
                        </WithStyles(TouchRipple)>
                      </button>
                    </ButtonBase>
                  </WithStyles(ButtonBase)>
                </IconButton>
              </WithStyles(IconButton)>
              <input aria-invalid={[undefined]} aria-required={[undefined]} autoComplete={[undefined]} autoFocus={[undefined]} className="MuiInput-input-15 RSFQuantitySelector-input-4 MuiInput-disabled-10" defaultValue={[undefined]} disabled={true} id={[undefined]} name={[undefined]} onBlur={[Function: value]} onChange={[Function: value]} onFocus={[Function: value]} onKeyDown={[undefined]} onKeyUp={[undefined]} placeholder={[undefined]} readOnly={[undefined]} required={[undefined]} rows={[undefined]} type="text" value={1} amp-bind="value=>undefined.quantity" />
              <WithStyles(IconButton) size="small" classes={{...}} onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 1) + 1) } })">
                <IconButton classes={{...}} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 1) + 1) } })" color="default" disabled={false}>
                  <WithStyles(ButtonBase) className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 1) + 1) } })">
                    <ButtonBase classes={{...}} className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 1) + 1) } })" disableRipple={false} tabIndex="0" type="button">
                      <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-26 MuiIconButton-root-20 RSFQuantitySelector-button-3" type="button" disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 1) + 1) } })">
                        <span className="MuiIconButton-label-25">
                          <pure(Add) classes={{...}}>
                            <Add classes={{...}}>
                              <WithStyles(SvgIcon) classes={{...}}>
                                <SvgIcon classes={{...}} color="inherit" viewBox="0 0 24 24">
                                  <svg className="MuiSvgIcon-root-29 RSFQuantitySelector-icon-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true">
                                    <g>
                                      <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />
                                    </g>
                                  </svg>
                                </SvgIcon>
                              </WithStyles(SvgIcon)>
                            </Add>
                          </pure(Add)>
                        </span>
                        <WithStyles(TouchRipple) innerRef={[Function: value]} center={true}>
                          <TouchRipple classes={{...}} center={true}>
                            <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-35" childFactory={[Function: childFactory]}>
                              <span className="MuiTouchRipple-root-35" />
                            </TransitionGroup>
                          </TouchRipple>
                        </WithStyles(TouchRipple)>
                      </button>
                    </ButtonBase>
                  </WithStyles(ButtonBase)>
                </IconButton>
              </WithStyles(IconButton)>
            </div>
          </Input>
        </WithStyles(Input)>
      </QuantitySelector>
    </WithStyles(QuantitySelector)>
  </inject-WithStyles(QuantitySelector)>
</Provider>
`;

exports[`QuantitySelector should accept a value 1`] = `
<Provider app={[Function: res]}>
  <inject-WithStyles(QuantitySelector) value={2}>
    <WithStyles(QuantitySelector) value={2} app={[Function: res]} ampStateId={[undefined]}>
      <QuantitySelector classes={{...}} value={2} app={[Function: res]} ampStateId={[undefined]} onChange={[Function]} minValue={1} maxValue={100}>
        <WithStyles(Input) startAdornment={{...}} endAdornment={{...}} onChange={[Function]} value={2} classes={{...}} inputProps={{...}} disabled={true}>
          <Input classes={{...}} startAdornment={{...}} endAdornment={{...}} onChange={[Function]} value={2} inputProps={{...}} disabled={true} disableUnderline={false} fullWidth={false} multiline={false} type="text">
            <div className="MuiInput-root-7 RSFQuantitySelector-root-1 MuiInput-disabled-10 MuiInput-underline-11 RSFQuantitySelector-underline-6">
              <WithStyles(IconButton) size="small" classes={{...}} onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 2) - 1) } })">
                <IconButton classes={{...}} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 2) - 1) } })" color="default" disabled={false}>
                  <WithStyles(ButtonBase) className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 2) - 1) } })">
                    <ButtonBase classes={{...}} className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 2) - 1) } })" disableRipple={false} tabIndex="0" type="button">
                      <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-26 MuiIconButton-root-20 RSFQuantitySelector-button-3" type="button" disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: max(1, (undefined.quantity || 2) - 1) } })">
                        <span className="MuiIconButton-label-25">
                          <pure(Remove) classes={{...}}>
                            <Remove classes={{...}}>
                              <WithStyles(SvgIcon) classes={{...}}>
                                <SvgIcon classes={{...}} color="inherit" viewBox="0 0 24 24">
                                  <svg className="MuiSvgIcon-root-29 RSFQuantitySelector-icon-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true">
                                    <g>
                                      <path d="M19 13H5v-2h14v2z" />
                                    </g>
                                  </svg>
                                </SvgIcon>
                              </WithStyles(SvgIcon)>
                            </Remove>
                          </pure(Remove)>
                        </span>
                        <WithStyles(TouchRipple) innerRef={[Function: value]} center={true}>
                          <TouchRipple classes={{...}} center={true}>
                            <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-35" childFactory={[Function: childFactory]}>
                              <span className="MuiTouchRipple-root-35" />
                            </TransitionGroup>
                          </TouchRipple>
                        </WithStyles(TouchRipple)>
                      </button>
                    </ButtonBase>
                  </WithStyles(ButtonBase)>
                </IconButton>
              </WithStyles(IconButton)>
              <input aria-invalid={[undefined]} aria-required={[undefined]} autoComplete={[undefined]} autoFocus={[undefined]} className="MuiInput-input-15 RSFQuantitySelector-input-4 MuiInput-disabled-10" defaultValue={[undefined]} disabled={true} id={[undefined]} name={[undefined]} onBlur={[Function: value]} onChange={[Function: value]} onFocus={[Function: value]} onKeyDown={[undefined]} onKeyUp={[undefined]} placeholder={[undefined]} readOnly={[undefined]} required={[undefined]} rows={[undefined]} type="text" value={2} amp-bind="value=>undefined.quantity" />
              <WithStyles(IconButton) size="small" classes={{...}} onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 2) + 1) } })">
                <IconButton classes={{...}} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 2) + 1) } })" color="default" disabled={false}>
                  <WithStyles(ButtonBase) className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 2) + 1) } })">
                    <ButtonBase classes={{...}} className="MuiIconButton-root-20 RSFQuantitySelector-button-3" centerRipple={true} focusRipple={true} disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 2) + 1) } })" disableRipple={false} tabIndex="0" type="button">
                      <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-26 MuiIconButton-root-20 RSFQuantitySelector-button-3" type="button" disabled={false} size="small" onClick={[Function: onClick]} on="tap:AMP.setState({ undefined: { quantity: min(100, (undefined.quantity || 2) + 1) } })">
                        <span className="MuiIconButton-label-25">
                          <pure(Add) classes={{...}}>
                            <Add classes={{...}}>
                              <WithStyles(SvgIcon) classes={{...}}>
                                <SvgIcon classes={{...}} color="inherit" viewBox="0 0 24 24">
                                  <svg className="MuiSvgIcon-root-29 RSFQuantitySelector-icon-2" focusable="false" viewBox="0 0 24 24" color={[undefined]} aria-hidden="true">
                                    <g>
                                      <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />
                                    </g>
                                  </svg>
                                </SvgIcon>
                              </WithStyles(SvgIcon)>
                            </Add>
                          </pure(Add)>
                        </span>
                        <WithStyles(TouchRipple) innerRef={[Function: value]} center={true}>
                          <TouchRipple classes={{...}} center={true}>
                            <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-35" childFactory={[Function: childFactory]}>
                              <span className="MuiTouchRipple-root-35" />
                            </TransitionGroup>
                          </TouchRipple>
                        </WithStyles(TouchRipple)>
                      </button>
                    </ButtonBase>
                  </WithStyles(ButtonBase)>
                </IconButton>
              </WithStyles(IconButton)>
            </div>
          </Input>
        </WithStyles(Input)>
      </QuantitySelector>
    </WithStyles(QuantitySelector)>
  </inject-WithStyles(QuantitySelector)>
</Provider>
`;
